zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -286,6 +293,91 @@ void rollbackPendingChanges(long zxid, Map<String, ChangeRecord>pendingChangeRec
         }
     }
 
+    /**
+     * check whether exceeded the quota.
+     *
+     * @param lastPrefix
+     *            the path of the node that is quotaed.
+     * @param bytesDiff
+     *            the diff to be added to number of bytes
+     * @param countDiff
+     *            the diff to be added to the count
+     */
+    private void checkQuota(String lastPrefix, long bytesDiff, int countDiff)
+            throws KeeperException.QuotaExceededException {
+        if (!enforeQuota) {
+            return;
+        }
+
+        LOG.debug("checkQuota: lastPrefix={}, bytesDiff={}, countDiff={}", lastPrefix, bytesDiff, countDiff);
+
+        if (lastPrefix == null || lastPrefix.length() == 0) {
+            return;
+        }
+        // now check if the counts match the quota
+        String limitNode = Quotas.limitPath(lastPrefix);
+        DataNode node = zks.getZKDatabase().getNode(limitNode);
+        StatsTrack limitStats;
+        if (node == null) {
+            // should not happen
+            LOG.error("Missing count node for quota " + limitNode);
+            return;
+        }
+        synchronized (node) {
+            limitStats = new StatsTrack(new String(node.data));
+        }
+        //check the quota
+        boolean checkCountQuota = (limitStats.getCount() > -1 || limitStats.getCountHardLimit() > -1);
+        boolean checkByteQuota = (limitStats.getBytes() > -1 || limitStats.getByteHardLimit() > -1);
+
+        if (!checkCountQuota && !checkByteQuota) {
+            return;
```

Comment:
Just thinking out loud: how often does that happen?  Should the administrator be alerted, somehow, that the clients are hitting an useless quota node?  Or am I missing an existing case and/or foreseen extension?

Reply:
I don't get the point :)